pkg/blockbuilder/blockbuilder.go

```diff
@@ -516,6 +521,19 @@ func (b *BlockBuilder) consumePartition(
 				// So the commit record should be the last record that was fully processed and not the
 				// first record that was not fully processed.
 				commitRec = lastRec
+
+				// The first record itself was not fully processed. Meaning the record before
+				// this is the commit point.
+				if commitRec == nil {
+					// When the commit record does not change from the previous commit, we get nil here.
+					commitRec = &kgo.Record{
+						Timestamp:   time.UnixMilli(pl.CommitRecTs),
+						Topic:       lastCommit.Topic,
+						Partition:   lastCommit.Partition,
+						LeaderEpoch: lastCommit.LeaderEpoch,
+						Offset:      lastCommit.At - 1,
+					}
```

Comment:
Nice catch. Just as an alternative, instead of building an artificial commit, can we `continue` here, leaving the `commitRec` and `lastRec` to be `nil`? In this case, the code below will skip doing any commit, and just rewind the offset to the first record (ref [code][1]).

[1]: https://github.com/grafana/mimir/blob/558ff24f6d4946aca7e94f64f2641ec32cbdb5b6/pkg/blockbuilder/blockbuilder.go#L553-L561

Reply:
From the PR description
> We do need to re-commit in order to update the "last seen" timestamp in the metadata.